### ISSUE TITLE
Add Bundle Adjustment Module

### ISSUE THREAD

**Issue by wang-chen** [pypose/pypose#120]
### 🚀 The feature, motivation and pitch

As stated in the title, a differential bundle adjustment module is required in many SLAM systems.
This module may require the following implemented functionalities.

- [LM optimizer](https://pypose.org/docs/main/generated/pypose.optim.LevenbergMarquardt/)
- [Model Jacobian Computation](https://pypose.org/docs/main/generated/pypose.optim.functional.modjac/)

**Comment by lizimo061** [comment]
Hi there, I wrote a small example based on [BAL dataset](https://grail.cs.washington.edu/projects/bal/), but I found it is extremely slow even on the smallest dataset(with the vectorize option off, CUDA out of memory reported if on), any good suggestion? 
`class BAGraph(nn.Module):

    def __init__(self, cameras, points, intrinsics):
        super().__init__()
        self.cameras = pp.Parameter(cameras)
        self.points = nn.Parameter(points)
        self.intrinsics = nn.Parameter(intrinsics)

    def forward(self, observations, camera_index, pt_index):
        pt_cam = self.cameras[camera_index, ...].Act(self.points[pt_index, ...])
        pt_image = -pt_cam[..., :2] / torch.unsqueeze(pt_cam[..., 2], 1)
        pt_norm = torch.linalg.vector_norm(pt_image, ord=2, dim=1)
        distortion_factors = 1 + self.intrinsics[camera_index, ..., 1] * pt_norm**2 + self.intrinsics[camera_index, ...,
                                                                                                      2] * pt_norm**4
        pt_pixel = torch.unsqueeze(self.intrinsics[camera_index, ..., 0], 1) * torch.unsqueeze(distortion_factors,
                                                                                               1) * pt_image
        error = pt_pixel - observations
        return error`

**Comment by wang-chen** [comment]
Hi, we are currently implementing sparse tensors, which will solve this problem. @zitongzhan

**Comment by zitongzhan** [comment]
BA is dependent on this issue: https://github.com/pypose/pypose/issues/207

**Comment by hxu296** [comment]
> Hi there, I wrote a small example based on [BAL dataset](https://grail.cs.washington.edu/projects/bal/), but I found it is extremely slow even on the smallest dataset(with the vectorize option off, CUDA out of memory reported if on), any good suggestion?
> 
> ```
> class BAGraph(nn.Module):
> 
>     def __init__(self, cameras, points, intrinsics):
>         super().__init__()
>         self.cameras = pp.Parameter(cameras)
>         self.points = nn.Parameter(points)
>         self.intrinsics = nn.Parameter(intrinsics)
> 
>     def forward(self, observations, camera_index, pt_index):
>         pt_cam = self.cameras[camera_index, ...].Act(self.points[pt_index, ...])
>         pt_image = -pt_cam[..., :2] / torch.unsqueeze(pt_cam[..., 2], 1)
>         pt_norm = torch.linalg.vector_norm(pt_image, ord=2, dim=1)
>         distortion_factors = 1 + self.intrinsics[camera_index, ..., 1] * pt_norm**2 + self.intrinsics[camera_index, ...,
>                                                                                                       2] * pt_norm**4
>         pt_pixel = torch.unsqueeze(self.intrinsics[camera_index, ..., 0], 1) * torch.unsqueeze(distortion_factors,
>                                                                                                1) * pt_image
>         error = pt_pixel - observations
>         return error
> ```

Hello, you can try an unofficial pypose-native BAL colab example [here]( https://colab.research.google.com/github/hxu296/pypose/blob/sparse/ba/examples/sparse/ba_example.ipynb). The smallest problem (problem-49-7776-pre) works fine on both CPU and GPU, and you can change the problem by using CLI parameters.

**Comment by chenjiajie9811** [comment]
Any update on the official implementation for the differentiable BA?

**Comment by zitongzhan** [comment]
> Any update on the official implementation for the differentiable BA?

We are preparing for it and tuning its final APIs.

**Comment by jiaming-ai** [comment]
hi, any expected date for releasing this feature?

**Comment by oliver-batchelor** [comment]
"[Bundle Adjustment in the Eager Mode](https://arxiv.org/abs/2409.12190)"  - is this an implementation for this feature in pypose?

**Comment by zitongzhan** [comment]
> "[Bundle Adjustment in the Eager Mode](https://arxiv.org/abs/2409.12190)" - is this an implementation for this feature in pypose?

Yes. It will be an implementation mimicking pypose API and sharing many components.

**Comment by oliver-batchelor** [comment]
> > "[Bundle Adjustment in the Eager Mode](https://arxiv.org/abs/2409.12190)" - is this an implementation for this feature in pypose?
> 
> Yes. It will be an implementation mimicking pypose API and sharing many components.

Very cool. I think it will get a lot of use.  

Does it exist as a branch or work-in-progress here or will it be a new library which adds this to pypose?
